openlibrary/catalog/add_book/match.py

```diff
@@ -53,13 +47,434 @@ def editions_match(candidate, existing):
     ):
         if existing.get(f):
             rec2[f] = existing[f]
+    # Transfer authors as Dicts str: str
     if existing.authors:
         rec2['authors'] = []
-        for a in existing.authors:
-            while a.type.key == '/type/redirect':
-                a = web.ctx.site.get(a.location)
-            if a.type.key == '/type/author':
-                assert a['name']
-                rec2['authors'].append({'name': a['name'], 'db_name': db_name(a)})
+    for a in existing.authors:
+        while a.type.key == '/type/redirect':
+            a = web.ctx.site.get(a.location)
+        if a.type.key == '/type/author':
+            author = {'name': a['name']}
+            if birth := a.get('birth_date'):
+                author['birth_date'] = birth
+            if death := a.get('death_date'):
+                author['death_date'] = death
+            rec2['authors'].append(author)
+    return threshold_match(rec, rec2, threshold)
+
+
+def normalize(s: str) -> str:
+    """
+    Normalizes a title for matching purposes, not display,
+    by lowercasing, unicode -> NFC,
+    stripping extra whitespace and punctuation, and replacing ampersands.
+    """
+
+    if isinstance(s, str):
+        # LATIN SMALL LETTER L WITH STROKE' (U+0142) -> 'l'
+        s = unicodedata.normalize('NFC', s.replace('\u0142', 'l'))
+    s = s.replace(' & ', ' and ')
+    # remove {mlrhring} and friends
+    # see http://www.loc.gov/marc/mnemonics.html
+    # s = re_brace.sub('', s)
+    s = re_whitespace_and_punct.sub(' ', s.lower())
+    s = re_normalize.sub('', s.strip())
+    return s
+
+
+def mk_norm(s: str) -> str:
+    """
+    Normalizes titles and strips ALL spaces and small words
+    to aid with string comparisons of two titles.
+    Used in comparing Work titles.
+
+    :param str s: A book title to normalize and strip.
+    :return: a lowercase string with no spaces, containing the main words of the title.
+    """
+    if m := re_brackets.match(s):
+        s = m.group(1)
+    norm = normalize(s).replace(' and ', '')
+    return strip_articles(norm).replace(' ', '')
+
+
+def strip_articles(s: str) -> str:
+    """
+    Strip articles for matching purposes.
+    TODO: Expand using
+    https://web.archive.org/web/20230320141510/https://www.loc.gov/marc/bibliographic/bdapndxf.html
+    or something sensible.
+    """
+    if s.lower().startswith('the '):
+        s = s[4:]
+    elif s.lower().startswith('a '):
+        s = s[2:]
+    return s
+
+
+def set_isbn_match(score):
+    isbn_match = score
```

Comment:
What is this? Looks unused. isbn_match looks like it should be a constant. Is it being used for any calculation?

Reply:
removed.